src/basilisp/core/protocols.lpy

```diff
@@ -0,0 +1,62 @@
+(ns basilisp.core.protocols)
+
+(defprotocol SeqReduce
+  "Protocol impl of reduce that's faster than first/next iteration. Called by reduce."
+  (seq-reduce-impl [s f init]))
+
+(extend-protocol SeqReduce
+  nil
+  (seq-reduce-impl [s f init] init)
+
+  python/object
```

Comment:
I haven't had time to really inspect the Clojure implementation, but I'm not sure what would be the purpose of attempting to iterate on any arbitrary object. Should we at least be constraining the iteration to something like `collections.abc.Iterable` or something we know we can get a seq from?

Reply:
Clojure implements on both `nil` and `Object` so that you can call `reduce` on anything, including an ArrayList or something (whatever can be coerced to a seq). If you specialize, you get a "Missing implementation for protocol CollReduce" error, whereas this way you get a "Can't call seq on a CustomObj" error, which I think is more informative.

I don't remember enough about the Python ecosystem to know if `collections.abc.Iterable` works for things that don't inherit from it but still implement `__next__` or whatever the correct dunder method is.

Reply:
> Clojure implements on both nil and Object so that you can call reduce on anything, including an ArrayList or something (whatever can be coerced to a seq). If you specialize, you get a "Missing implementation for protocol CollReduce" error, whereas this way you get a "Can't call seq on a CustomObj" error, which I think is more informative.

I'm wavering on agreement here. I think newer users may find that implementation detail error message somewhat confusing, though I think neither message is ideal.

> I don't remember enough about the Python ecosystem to know if collections.abc.Iterable works for things that don't inherit from it but still implement __next__ or whatever the correct dunder method is.

For now I think it's fine to just use what you have, but I do believe now due to the implementation of virtual subclasses in the Python VM that we can use any of the defined Python abstract base classes even if the desired class does not directly inherit from the ABC.